### ISSUE TITLE
InitRegionData() and friends allocate data unnecessarily

### ISSUE THREAD

**Issue by randomascii** [google/libaddressinput#144]
TL;DR InitRegionData and friends do thousands of allocations and generate ~10 KB more code than needed.

InitRegionData() takes pairs of string constants, turns them into pairs of std::string objects, and then inserts those into an std::map<std::string, std::string>. This potentially requires three allocations for each pair, although due to the small-string optimization it is probably only two. There are currently 248 pairs, so 496 allocations for this function. These allocations could be avoided if InitRegionData was replaced with something like constexpr std::pair<const char* const, const char* const> region_data[]. This requires zero allocations and keeps all of the string data in constant memory. A binary search over such an array will be faster than searching over a set because of better memory locality.

InitRegionData() also generates significant code - 9370 bytes for Chrome's 64-bit canary build - and requires hooking up a destructor to clear out the data.

If InitRegionData were redone as a constexpr array then the functions which depend on it could also be changed to be constexpr data, and even more allocations would be avoided.

This isn't a life changing issue, but since the code can be written to avoid allocations and since this will improve performance and reduce code size it seems worthwhile.

In preparation for a change like this some of the loops could be rewritten to be more tolerant of different container types. For instance, instead of this loop:

  for (std::map<std::string, std::string>::const_iterator
       it = GetAllRegionData().begin(); it != GetAllRegionData().end(); ++it) {

something like this would work:

  for (const auto& it : GetAllRegionData()) {

**Comment by randomascii** [comment]
P.S. This was discovered when examining a list of atexit destructors generated by Chrome's 64-bit build on Windows. The following destructors were noticed:

obj/third_party/libaddressinput/util/region_data_constants.obj: `i18n::addressinput::`anonymous namespace'::GetAllRegionData'::`2'::`dynamic atexit destructor for 'kRegionData''
obj/third_party/libaddressinput/util/region_data_constants.obj: `i18n::addressinput::RegionDataConstants::GetDefaultRegionData'::`2'::`dynamic atexit destructor for 'kDefaultRegionData''
obj/third_party/libaddressinput/util/region_data_constants.obj: `i18n::addressinput::RegionDataConstants::GetMaxLookupKeyDepth'::`2'::`dynamic atexit destructor for 'kMaxDepth''
obj/third_party/libaddressinput/util/region_data_constants.obj: `i18n::addressinput::RegionDataConstants::GetRegionCodes'::`2'::`dynamic atexit destructor for 'kRegionCodes''
obj/third_party/libaddressinput/util/region_data_constants.obj: `i18n::addressinput::RegionDataConstants::GetRegionData'::`2'::`dynamic atexit destructor for 'kEmptyString''
obj/third_party/libaddressinput/util/region_data_constants.obj: `i18n::addressinput::RegionDataConstants::IsSupported'::`2'::`dynamic atexit destructor for 'kRegionCodes''

**Comment by roubert** [comment]
The recent switch of libaddressinput to C++11 (commit a1dfd1b) should make this rather straightforward to do.

**Comment by madmath** [comment]
This is actionable now, right?

**Comment by roubert** [comment]
Yes, since the upgrade to C++11 last year (see my comment from October) it should be quite straightforward to rewrite `GetAllRegionData()` in some way that doesn't do that many allocations.

The code was originially written the way it was just because it was easy and obvious, without any regard to performance. Now when it's known that performance is an issue here and how to measure it, it should be easy enough to optimize.

Just note that the `region_data_constants.cc` source file is autogenerated, so any changes made to it must be propagated to the generator code (or else they'd be overwritten the next time the file is regenerated).

**Comment by rsolomakhin** [comment]
Should have been fixed in https://github.com/google/libaddressinput/pull/167
